#**LAB-01 : Create a program to build and train an Artificial Neural Network**

In [ ]:
import pandas as pd
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import FunctionTransformer

In [ ]:
# Load the dataset
file_path = '/content/Titaanic Datasetforlogisticregression .csv'
df = pd.read_csv(file_path)

In [ ]:
df

Survived  Pclass     Sex   Age  Siblings/Spouses Aboard  \
0           0       3    male  22.0                        1   
1           1       1  female  38.0                        1   
2           1       3  female  26.0                        0   
3           1       1  female  35.0                        1   
4           0       3    male  35.0                        0   
..        ...     ...     ...   ...                      ...   
882         0       2    male  27.0                        0   
883         1       1  female  19.0                        0   
884         0       3  female   7.0                        1   
885         1       1    male  26.0                        0   
886         0       3    male  32.0                        0   

     Parents/Children Aboard     Fare  
0                          0   7.2500  
1                          0  71.2833  
2                          0   7.9250  
3                          0  53.1000  
4                          0   8.0500  
..                       ...      ...  
882                        0  13.0000  
883                        0  30.0000  
884                        2  23.4500  
885                        0  30.0000  
886                        0   7.7500  

[887 rows x 7 columns]

In [ ]:
df['Age'].fillna(df['Age'].median(), inplace=True)
df['Fare'].fillna(df['Fare'].median(), inplace=True)

In [ ]:
# Feature Engineering
df['Family_Size'] = df['Siblings/Spouses Aboard'] + df['Parents/Children Aboard']
df.drop(['Siblings/Spouses Aboard', 'Parents/Children Aboard'], axis=1, inplace=True)

In [ ]:
# One-Hot Encoding for categorical variables
df = pd.get_dummies(df, columns=['Sex', 'Pclass'], drop_first=True)

In [ ]:
# Split the data into features and target
X = df.drop('Survived', axis=1)
y = df['Survived']

In [ ]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
# Standardize the data
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [ ]:
# Build the ANN model
model = Sequential()
model.add(Dense(units=128, activation='relu', input_shape=(X_train.shape[1],)))
model.add(Dropout(0.5))
model.add(Dense(units=64, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(units=64, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(units=64, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(units=32, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(units=16, activation='relu'))
model.add(Dense(units=1, activation='sigmoid'))

In [ ]:
# Compile the model
model.compile(optimizer=Adam(learning_rate=0.01), loss='binary_crossentropy', metrics=['accuracy'])

In [ ]:
# Train the model
history = model.fit(X_train, y_train, epochs=50, batch_size=50, validation_split=0.2, verbose=1)

Epoch 1/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 4s 43ms/step - accuracy: 0.5977 - loss: 0.6769 - val_accuracy: 0.6197 - val_loss: 0.5942
Epoch 2/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.6426 - loss: 0.6128 - val_accuracy: 0.7958 - val_loss: 0.5951
Epoch 3/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7047 - loss: 0.5771 - val_accuracy: 0.7887 - val_loss: 0.4979
Epoch 4/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8144 - loss: 0.5050 - val_accuracy: 0.8028 - val_loss: 0.4217
Epoch 5/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7768 - loss: 0.5297 - val_accuracy: 0.8169 - val_loss: 0.5101
Epoch 6/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8041 - loss: 0.4970 - val_accuracy: 0.8380 - val_loss: 0.4360
Epoch 7/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7850 - loss: 0.5306 - val_accuracy: 0.8380 - val_loss: 0.4170
Epoch 8/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8202 - loss: 0.4670 - val_accuracy: 0.8521 - val_los

In [ ]:
# Evaluate the model
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print(f'Test Accuracy: {test_accuracy:.4f}')

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7529 - loss: 0.5883 
Test Accuracy: 0.7865


**Inference:** The model achieved a test accuracy of approximately 78.65%. This means that the model correctly classified about 78.65% of the test data samples.
The test loss is 0.5883 . The loss value represents how well or poorly the model's predictions align with the actual labels.
Lower loss indicates better performance, although the acceptable range of loss values depends on the specific problem and dataset.